pkg/controller/syncmachineset/syncmachineset_controller.go

```diff
@@ -0,0 +1,548 @@
+	// Load the kubeconfig secret for communicating with the remote cluster:
+
+	// Step 1: Get a cluster from the machineset
+	cluster, err := controller.ClusterForMachineSet(machineSet, c.clusterLister)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve cluster for machineset")
+	}
+
+	// Step 2: Get the secret ref from the cluster
+	//
+	// TODO: The cluster controller should watch secrets until it
+	// sees a kubeconfig secret named after the cluster and then
+	// update cluster status with a reference to that
+	// secret. Until then we'll look for a
+	// "clustername-kubeconfig" secret.
+	//
+	// clusterConfig := cluster.Status.AdminKubeconfig
+	// if clusterConfig == nil {
+	// 	return nil, fmt.Errorf("unable to retrieve AdminKubeConfig from cluster status")
+	// }
+	secretName := cluster.Name + "-kubeconfig"
+
+	// Step 3: Retrieve secret
+	secret, err := c.kubeClient.CoreV1().Secrets(machineSet.Namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	secretData := secret.Data["kubeconfig"]
+
+	// Step 4: Generate config from secret data
+	config, err := clientcmd.Load(secretData)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{})
+	restConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Step 5: Return client for remote cluster
+	return clusterapiclient.NewForConfig(restConfig)
+}
+
+// syncMachineSet will sync the machine set with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncMachineSet(queuedMS queuedMachineSet) error {
+	startTime := time.Now()
+	key := queuedMS.key
+	c.logger.WithField("key", key).Debug("syncing machineset")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing machineset")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	machineSet, err := c.machineSetsLister.MachineSets(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("machineset has been deleted")
+		// TODO: ensure deleted in remote cluster. If so, how
+		// are we going to figure out what cluster the machine
+		// set was from to load the correct kubeconfig secret?
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Master machinesets are managed by the CO cluster and are
+	// not synced with the remote cluster. When syncing a master
+	// machineset, we will sync the cluster that the master
+	// machineset belongs to rather than the master machineset
+	// itself.
+	if isMasterMachineSet(machineSet) {
+		return c.syncMasterMachineSet(machineSet)
+	}
+
+	return c.syncComputeMachineSet(machineSet)
+}
+
+// Creates the cluster in the remote.
+func (c *Controller) syncMasterMachineSet(ms *cov1.MachineSet) error {
```

Comment:
This should really be a controller that is syncing on the cluster instead of the machineset, since it is the cluster that is the operative component. It's fine for now though given that this is temporary.